### PR TITLE
"Device Editor" enabled by default

### LINKED ISSUE #2884: "Open Editor" enabled by default when in Developer Mode
### Description

From discussion with @Steve-Mcl and @ZJvandeWeg (https://flowforgeworkspace.slack.com/archives/C03FAUBPPJ7/p1696424993218099)

If Editor Access is not enabled at that point, we enable it ourselves, tell them it's opening connection, etc. and then open the Editor.

Existing UX around opening/closing tunnel still exists.

### Have you provided an initial effort estimate for this issue?

I have provided an initial effort estimate

### PR DESCRIPTION
**Pull request by Steve-Mcl**
closes #2884  

## Description

Enabled the "Open Editor" button by default (when in developer mode)

To support this, a dialog was added on the page that clearly indicates the editor is connecting (since it takes some time over the internet) and provides a simple and quick means of cancelling the request (which also closes the tunnel) which is necessary if the user is not viewing the "developer mode" tab where the tunnel can be disabled.

Additionally, during demonstration, it was noted the tunnel can become disconnected (and irretrievable) if the device agent is restarted AFTER the tunnel is establised.  A page refresh and manually visiting the "developer mode" tab was necessary to both "know this" and "recover this".  Changes were made to simplify recovery:
1. Move the polling timer from `pages/device/Overview.vue` to its parent  `pages/device/index.vue` - this means the tunnel status is updated and the FE can know if the tunnel was lost. Meaning the "open device" button have attempt to re-establish the tunnel.
2. In the route `/api/v1/devices/:deviceId/editor` I check to see if the tunnel is enabled but NOT connected. If so, attempt to recreate the tunnel. Since this is a deliberate user action (and not just a remote device editor refresh) it feels like the right thing to do.

Here is a short video demonstrating the above:

https://github.com/FlowFuse/flowfuse/assets/44235289/bc4b1cd9-fbbf-4576-83b3-b02be519e4fc

## Related Issue(s)

#2884 

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [ ] Suitable unit/system level tests have been added and they pass 
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Backport needed? -> add the `backport` label
 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Comment by Steve-Mcl** [comment]
@Pezmc added as additional reviewer following merge conflict with #3052

File: frontend/src/pages/device/index.vue [hunk 1]
```diff
@@ -57,10 +51,38 @@
...
+                <router-view :instance="device.instance" :closingTunnel="closingTunnel" :openingTunnel="openingTunnel" :device="device" @device-updated="loadDevice" @close-tunnel="closeTunnel" @open-tunnel="openTunnel" @device-refresh="loadDevice" @assign-device="openAssignmentDialog" />
             </div>
         </div>
         <!-- Dialogs -->
+        <!-- device tunnel connecting -->
+        <ff-dialog ref="dialog" data-el="establish-device-tunnel-dialog" header="Preparing the connection...">
+            <template #default>
+                <div class="flex flex-col ml-6 mr-6">
+                    <div class="mb-4">
+                        <p>Connecting. Please wait...</p>
```
Comment (knolleary) [inline frontend/src/pages/device/index.vue:63]:
```suggestion
                        <p>Connecting to the device</p>
```

Not going to get too hung up on the precise words - just a suggestion here to be more explicit that we're connecting to the device.

**Review by knolleary — APPROVED** [review]
Overall approved. A couple messages I'd suggest rewording slightly.

File: frontend/src/pages/device/index.vue [hunk 2]
```diff
@@ -259,7 +307,74 @@ export default {
...
+                                this.openEditor()
+                            }
+                            this.$refs.dialog.close()
+                        }, 500)
+                    } finally {
+                        this.openingTunnel = false
+                    }
+                }
+            } else {
+                Alerts.emit('The device must be in "running" state to access the editor', 'warning', 7500)
```
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply frontend/src/pages/device/index.vue:337]: Hmmm. Not sure TBH. 

It could be any number of reasons - far more than would "look nice" in a toast.

For example: The device might be off-line due to network, power, busy starting up, removed, and so on. 

How about:

```suggestion
                Alerts.emit('Unable to establish a connection to the device. Please ensure it is powered on, connected to the network and is in the "running" state before trying again', 'warning', 7500)
```

**Comment by Steve-Mcl** [comment]
> @Pezmc added as additional reviewer following merge conflict with #3052

@Pezmc ^

**Review by Pezmc — APPROVED** [review]

File: frontend/src/pages/device/index.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply frontend/src/pages/device/index.vue:337]: @knolleary do you have a cleaner, more succinct suggestion pls? ^

[the remark this answers is not in the document]
Comment (knolleary) [reply frontend/src/pages/device/index.vue:337]: Something like:

> Unable to establish a connection to the device. Please check it is connected and try again

[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply frontend/src/pages/device/index.vue:337]:
```suggestion
                Alerts.emit('Unable to establish a connection to the device. Please check it is connected and running then try again', 'warning', 7500)
```

Slight variation for consideration.  Since the logic test that leads to this toast is that the device status is NOT "running" (i.e. it could be starting up or crashed or some other state).

Thoughts please @knolleary

**Comment by ZJvandeWeg** [comment]
@knolleary @Steve-Mcl This PR is missing items from the definition of done. Like docs.

Furthermore, will you write a changelog entry?

**Comment by ZJvandeWeg** [comment]
Also, there are no tests added?

**Comment by Steve-Mcl** [comment]
@ZJvandeWeg 

> the definition of done. Like docs.

ZJ, I did scan the docs and there was (IMO) no need for an update.  But now you mention this I am starting to think there is a place that needs revising?  I will have another scan of the docs and if necessary, raise a follow up issue. 


> Furthermore, will you write a changelog entry? 

Sorry if I am missing something here but where do you want this ZJ? The change log is auto generated by the release process.

> here are no tests added

Yes, I slipped up on this.

Please accept this tenuous explanation ;) 
When I demoed the feature in eng meeting 2 weeks ago, the changes were 100% frontend (mostly cosmetic) and existing tests were suitable.  We encountered a bug (an existing bug) where by if a device-agent is restarted, the tunnel is closed and not recoverable unless you 1. refresh the page, 2. close the (broken) tunnel, 3. re-enable the tunnel.  It was observed and commented upon and suggested this be better handled - which is when I went back to the code and did the commit [improve "not connected" issue when agent restarted](https://github.com/FlowFuse/flowfuse/pull/3061/commits/ec9cd6d27a57458a6c80328166b78d6b3e337291). At this retrospective point, it did completely slip my mind that I should have added a test for this. So definitely "my bad" here. 

I'll raise a follow up issue and point it at [96a2dbf](https://github.com/FlowFuse/flowfuse/pull/3061/commits/96a2dbfcb0520c15abd622a289f7cb88b17ed94f)



---


In summary (and for my own tracking) @Steve-Mcl:
1. Re-check docs, raise issue & fix up if necessary
2. Add follow up issue to add tests

These will be done today (if time), otherwise tomorrow.
